x/appchain/coordinator/keeper/slash.go

```diff
@@ -3,12 +3,95 @@
 	types "github.com/ExocoreNetwork/exocore/x/appchain/coordinator/types"
 	sdk "github.com/cosmos/cosmos-sdk/types"
 )
 
-// TODO: this file should be in the x/avs keeper instead.
+// ValidateSlashPacket validates a slashing packet. It checks that
+// (1) the valset update id maps back to an Exocore height, and
+// (2) the validator (cons) address maps back to an operator account address.
+// The caller must perform stateless validation by themselves.
+func (k Keeper) ValidateSlashPacket(
+	ctx sdk.Context, chainID string, data commontypes.SlashPacketData,
+) error {
+	// the validator set is generated at each epoch and following each slash. even if the set
+	// does not change at each epoch, we send an empty update anyway since, otherwise, the
+	// subscriber could time out the coordinator.
+	if k.GetHeightForChainVscID(ctx, chainID, data.ValsetUpdateID) == 0 {
+		return commontypes.ErrInvalidPacketData.Wrapf(
+			"invalid chainID %s valsetUpdateID %d", chainID, data.ValsetUpdateID,
+		)
+	}
+	// the second step is to find the operator account address against the consensus address
+	// of the validator. if the operator is not found, the slashing packet is invalid.
+	// note that this works even if the operator changes their consensus key, as long as the
+	// key hasn't yet been pruned from the operator module.
+	if found, _ := k.operatorKeeper.GetOperatorAddressForChainIDAndConsAddr(
+		ctx, chainID, sdk.ConsAddress(data.Validator.Address),
+	); !found {
+		// don't bech32 encode it in the error since the appchain may have a different prefix
+		return commontypes.ErrInvalidPacketData.Wrapf(
+			"operator not found %x", data.Validator.Address,
+		)
+	}
+	return nil
+}
+
+// HandleSlashPacket handles a slashing packet. The caller must ensure that the slashing packet
+// is valid before calling this function. The function forwards the slashing request to the
+// operator module, which will trigger a slashing hook and thus a validator set update containing
+// the slashing acknowledgment.
+func (k Keeper) HandleSlashPacket(ctx sdk.Context, chainID string, data commontypes.SlashPacketData) {
+	consAddress := sdk.ConsAddress(data.Validator.Address)
+	// never 0, since already validated
+	height := k.GetHeightForChainVscID(ctx, chainID, data.ValsetUpdateID)
+	// guaranteed to exist, since already validated
+	_, operatorAccAddress := k.operatorKeeper.GetOperatorAddressForChainIDAndConsAddr(
+		ctx, chainID, consAddress,
+	)
+	slashProportion := k.GetSubSlashFractionDowntime(ctx, chainID)
+	// #nosec G703 // already validated
+	slashProportionDecimal, _ := sdk.NewDecFromStr(slashProportion)
+	jailDuration := k.GetSubDowntimeJailDuration(ctx, chainID)
+	chainIDWithoutRevision := utils.ChainIDWithoutRevision(chainID)
+	_, avsAddress := k.avsKeeper.IsAVSByChainID(ctx, chainIDWithoutRevision)
+	// the slashing hook should trigger a validator set update for all affected AVSs. since the `chainID` is one of them
```

Comment:
Not necessary since the packet must be received over a channel belonging to a chain known to the coordinator module, which mandatorily registers the AVS.